Kernel/Syscalls/uname.cpp

```diff
@@ -13,19 +16,67 @@ ErrorOr<FlatPtr> Process::sys$uname(Userspace<utsname*> user_buf)
     VERIFY_NO_PROCESS_BIG_LOCK(this);
     TRY(require_promise(Pledge::stdio));
 
-    utsname buf {};
-    memcpy(buf.sysname, "SerenityOS", 11);
-    memcpy(buf.release, "1.0-dev", 8);
-    memcpy(buf.version, "FIXME", 6);
+    utsname buf
+    {
+        "SerenityOS",
+            "", // Hostname, filled in below.
+            "", // "Release" (1.0-dev), filled in below.
+            "", // "Revision" (git commit hash), filled in below.
 #if ARCH(I386)
-    memcpy(buf.machine, "i686", 5);
+            "i686",
 #elif ARCH(X86_64)
-    memcpy(buf.machine, "x86_64", 7);
+            "x86_64",
 #elif ARCH(AARCH64)
-    memcpy(buf.machine, "AArch64", 7);
+            "AArch64",
 #else
 #    error Unknown architecture
 #endif
+    };
+
+    // The full version is found in /res/version.ini, so let's read that like the user would.
+    auto description = TRY(VirtualFileSystem::the().open(credentials(), "/res/version.ini"sv, O_RDONLY, 0000, current_directory()));
+    ScopeGuard guard([&]() { (void)description->close(); });
+    auto raw_contents = TRY(description->read_entire_file());
+    auto contents = StringView(raw_contents->bytes());
+
+    // FIXME: This INI parsing is super crude but we can't use Core::ConfigFile. As long as our version.ini doesn't do weird things, it should work.
+    // Find a way of reusing code across Kernel and Userland.
+    auto maybe_section_location = contents.find("[Version]\n"sv);
+    if (maybe_section_location.has_value()) {
+        auto section_location = maybe_section_location.value() + 10;
+        auto major_version_location = contents.find("Major="sv, section_location);
+        auto minor_version_location = contents.find("Minor="sv, section_location);
+        auto git_hash_location = contents.find("Git="sv, section_location);
+
+        StringBuilder version_builder;
+
+        auto append_version_from_config = [&](auto maybe_location, auto key_length) {
+            if (maybe_location.has_value()) {
+                size_t i = maybe_location.value() + key_length;
+                while (i < contents.length() && contents[i] != '\n')
+                    version_builder.append(contents[i++]);
+            } else {
+                version_builder.append('0');
+            }
+        };
+
+        append_version_from_config(major_version_location, 6);
+        version_builder.append('.');
+        append_version_from_config(minor_version_location, 6);
+        version_builder.append("-dev"sv);
+
+        AK::TypedTransfer<u8>::copy(reinterpret_cast<u8*>(buf.release), version_builder.to_byte_buffer().data(), min(version_builder.length(), UTSNAME_ENTRY_LEN - 1));
+
+        // Reuse the string builder for the version, i.e. the git commit hash.
+        version_builder.clear();
+
+        if (git_hash_location.has_value())
+            append_version_from_config(git_hash_location, 4);
+        else
+            version_builder.append("FIXME"sv);
```

Comment:
Let's use "unknown" here

Reply:
That would be a change in behaviour, I intentionally didn't want that if the git hash wasn't available, but sure.